Create fab-files with a Makefile

hzeller opened the pull request:
Hi Jason,
In the context of the BeagleBone Blue reporting, I stumbled upon your Octavo-based PocketBone project. To make things simple to fab, I added a Makefile that creates a zip file with all of them.

Also, I personally like if the fabbed output contains a git hash and timestamp so that it is easy to find the right version when your desk is full of various stages of the project :). I think that can be useful here as well. So added that by adding a string on the back silk screen `%%gitversion%%` and a sed script in the Makefile that replaces that with the git log version of the pcb file.

(The standard Protel names for the inner layers are created with suffix *.g2 and *.g3. For some reason, OshPark does not recognize these but wants *.g2l and *.g3l. I opened a ticket with them as they promise that they understand the Protel names generated by KiCAD).

This requires, that the Python script extension for KiCAD are installed on the machine.

Cheers,
  Henner (the BeagleG guy).

o Add Python script that can generate fabbing files from the command …
o Add Makefile that makes use of it: just calling 'make' will create
  a zip file with the necessary layers.
o Send kicad_pcb through a sed script to change %%gitversion%% with
  the git hash and timestamp of the pcb file.
o Add a %%gitversion%% string on the back silkscreen, so that it
  is simple to identify the particular version of a fabbed board.

hzeller committed:
o Add Python script that can generate fabbing files from the command line.
o Add Makefile that makes use of it: just calling 'make' will create
  a zip file with the necessary layers.
o Send kicad_pcb through a sed script to change %%gitversion%% with
  the git hash and timestamp of the pcb file.
o Add a %%gitversion%% string on the back silkscreen, so that it
  is simple to identify the particular version of a fabbed board.

jadonk commented on pocketbone-kicad.kicad_pcb:
```diff
@@ -1,11 +1,11 @@
     (links 202)
     (no_connects 0)
-    (area 105.891359 94.855159 161.041361 130.005161)
+    (area 103.93175 93.91396 162.501581 132.574809)
```
Why is the board outline adjusted?

jadonk commented on pocketbone-kicad.kicad_pcb:
```diff
@@ -490,6 +489,8 @@
     (via_drill 0.3302)
     (uvia_dia 0.3)
     (uvia_drill 0.1)
+    (diff_pair_gap 0.25)
```
Looks handy. Is this a design rule check?

jadonk commented on pocketbone-kicad.kicad_pcb:
```diff
@@ -1871,7 +1872,7 @@
     (path /5838D871)
     (attr smd)
-    (fp_text reference C2 (at 1.609 0.0508 360) (layer F.SilkS)
+    (fp_text reference C2 (at 1.609 0.0508) (layer F.SilkS)
```
Lots of these. don't know if they are harmless.

jadonk commented on pocketbone-kicad.kicad_pcb:
```diff
@@ -4608,6 +4609,9 @@
       (xy 4.817536 3.229429) (xy 4.839415 3.147786)) (layer B.SilkS) (width 0.01))
   )
 
+  (gr_text %%gitversion%% (at 123.825 112.395 90) (layer B.SilkS) (tstamp 58CDFF39)
```
This seems to be the main point of the edit. likely needs to be isolated in the patch.

jadonk replied:
BTW, I love the idea of the silkscreen indicating the PCB revision. Was that one of the main thrusts?

hzeller replied:
This is now isolated, yep.

jadonk reviewed (changes requested):
Please annotate the rest of the changes in the source files.

hzeller commented:
I think most of these changes are due to opening it with one version of kicad and saving it with another.
The changes all look benign: for instance 
```
  (fp_text reference C2 (at 1.609 0.0508 360) (layer F.SilkS)
```
turns something by 360 degrees, and the newer KiCad decided that this is the same as not turning it by zero degrees
```
   (fp_text reference C2 (at 1.609 0.0508) (layer F.SilkS)
```
There are other cases in which multiple of 360 degrees are normalized.

Anyway, if you want I can just manually add the text to the pcb file, so that the changes are minimal.

hzeller committed:
o Limit changes in PCB file to only include the %%gitversion%% change.

hzeller committed:
o OshPark only successfully recongizes the file names if
  EdgeCuts is named Edge_Cuts, so renmae that file.

hzeller commented:
Alright, I have now updated the change by
  * Only including the one text addition in the PCB file needed for the %%gitversion%% string.
  * Updated the Python script to work reliably with OshPark (In the ticket they told me that it would work well if the edge file would be named `*Edge_Cuts*` instead of `*EdgeCuts*`.

The motivation for this change is not only the %%gitversion%% thing, but also that it is possible to have a one-stop way to create the gerber files from the raw files. It is very annoying to have to open a UI just to create gerber files. With the script, everything is already contained to make it work.

Since these are now three accumulated submits, I suggest to squash the pull request.

jadonk commented:
@mwelling I'm trying to keep this repo up-to-date. Do you also plan to merge this script? Shall I merge and create a pull request back to you?

mwelling commented:
@jadonk Since the pull request is on your repository, I think you should merge and I can pull the changes.

hzeller commented:
Is there anything else needed from me ?
